crates/nu-command/src/filesystem/cp.rs

```diff
@@ -345,29 +377,99 @@ fn interactive_copy(
-                ErrorKind::PermissionDenied => ShellError::PermissionDeniedError(message, span),
-                ErrorKind::Interrupted => ShellError::IOInterrupted(message, span),
-                ErrorKind::OutOfMemory => ShellError::OutOfMemoryError(message, span),
-                // TODO: handle ExecutableFileBusy etc. when io_error_more is stabilized
-                // https://github.com/rust-lang/rust/issues/86442
-                _ => ShellError::IOErrorSpanned(message, span),
-            };
-
-            Value::Error { error: shell_error }
-        }
+        Err(e) => error_return(e, src, span),
+    }
+}
+
+// `copy_file_with_progressbar` uses `read` and `write` to copy a file. There is another function called `copy_file`
+// which uses `std::fs::copy` instead which is faster but can't gt the process of the copy. If you make any changes
+// in this function try to match the changes in `copy_file`
+fn copy_file_with_progressbar(src: PathBuf, dst: PathBuf, span: Span) -> Value {
+    let mut bytes_processed: u64 = 0;
+    let bytes_processed_p = &mut bytes_processed;
+    let mut process_failed: Option<std::io::Error> = None;
+    let process_failed_p = &mut process_failed;
+
+    let file_in = match std::fs::File::open(&src) {
+        Ok(file) => file,
+        Err(error) => return error_return(error, src, span),
+    };
+
+    let file_size = match file_in.metadata() {
+        Ok(metadata) => Some(metadata.len()),
+        _ => None,
+    };
+
+    let bar = progress_bar::NuProgressBar::new(file_size);
+    let mut bar_clone = bar.clone();
+
+    let file_out = match std::fs::File::create(&dst) {
+        Ok(file) => file,
+        Err(error) => return error_return(error, src, span),
+    };
+    let mut buffer = [0u8; 8192];
+    let mut buf_reader = BufReader::new(file_in);
+    let mut buf_writer = BufWriter::new(file_out);
+
+    loop {
+        // Read src file
+        match buf_reader.read(&mut buffer) {
+            // src file read successfully
+            Ok(bytes_read) => {
+                // Write dst file
+                match buf_writer.write(&buffer[..bytes_read]) {
+                    // dst file written successfully
+                    Ok(bytes_written) => {
+                        // Update the total amount of bytes that has been saved and then print the progress bar
+                        *bytes_processed_p += bytes_written as u64;
+                        bar_clone.update_bar(*bytes_processed_p);
+
+                        // the last block of bytes is going to be lower than the buffer size
+                        // let's break the loop once we write the last block
+                        if bytes_read < buffer.len() {
+                            break;
+                        }
+                    }
+                    Err(e) => {
+                        // There was a problem writing the dst file
+                        *process_failed_p = Some(e);
```

Comment:
I think you forgot to `break` out of the loop when an error occurs, no?

Reply:
omg so embarrassed.